Add youtube hls streaming wip

mv2731 opened the pull request:
This pull request includes the changes made to support HLS ingestion streaming to YouTube. I added an obs_ouptut_info struct and registered it in order to make this possible and then wrote the corresponding methods for the struct. Additionally, I added parameters so that start and init functions could recognize the output format as HLS.

mv2731 committed:
plugins: Add Youtube-HLS to the Service dropdown

This commit changes the name of the option called Youtube
/ Youtube Gaming to Youtube-RTMP to be more specific
and adds a Youtube-HLS option in the Stream settings
Service dropdown.

mv2731 committed:
Revert "plugins: Add Youtube-HLS to the Service dropdown"

This reverts commit 92e35faf89154d83c772ca3e68d47a9561c2f1f7.

mv2731 committed:
plugins: Add Youtube-HLS to the Service dropdown

This commit changes the name of the option called Youtube
/ Youtube Gaming to Youtube-RTMP to be more specific
and adds a Youtube-HLS option in the Stream settings
Service dropdown.

mv2731 committed:
plugins: Fix URLs of Youtube-HLS servers

mv2731 committed:
plugins: Register and initialize ffmpeg_hls_muxer

This commit
	-registers ffmeg_hls_muxer as an output in obs
	-declares and initializes an obs_output_info struct
		for ffmpeg_hls_muxer
	-defines necessary functions for this struct's
		initialization

for the purpose of supporting HLS ingestion from the Service menu
under Stream settings.

mv2731 committed:
UI: Use || for "or-ing" service names

mv2731 committed:
plugins: Added spaces to Youtube service names

mv2731 committed:
plugins: Add in support to get stream_key for HLS

mv2731 committed:
plugins: Fix errors in obs compilation and run

mv2731 committed:
plugins: Add HLS ingestion support

Edit the start function of ffmpeg_hls_muxer to enable
HLS ingestion.

mv2731 committed:
plugins: Add log statements

Add log statements to obs-ffmpeg-mux.c and ffmpeg-mux.c
to better understand why ingestion isn't yet working.

mv2731 committed:
plugins: Add format parameter to indicate hls

Add format parameter so when the build_command_line function
is called an hls option can be specified and its value will
be stored in the format field of main_params.

mv2731 committed:
plugins: Enable HLS ingestion to YouTube

Fix function called ffmpeg_hls_mux_start and add a format parameter
to struct ffmpeg_muxer to indicate whether the format is hls or not.
Check whether this option is set in ffmpeg_mux_init_context to determine
the output_format.

mv2731 committed:
plugins: Fix segfault after second start

Fix segfault that occurs after starting a previously
stopped HLS livestream. Fix tokenizing of stream url
by copying it into a buffer.

mv2731 committed:
plugins: Modify string manipulation of path url

Modify string manipulation of path url for HLS ingestion
to use dstr methods, resulting in simpler more readable
code.

mv2731 committed:
plugins: Clean up git diffs before making PR

Remove white-spaces and unintended changes.

mv2731 committed:
tmp

mv2731 committed:
plugins: Re-added newline at end of package.json

mv2731 committed:
plugins: Fix the memory leak

Free the dstr used to fill in the HLS streaming
path with the stream key so as to stop a memory leak
from occurring.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -354,6 +363,51 @@ static bool ffmpeg_mux_start(void *data)
+	stream->total_bytes = 0;
+	obs_output_begin_data_capture(stream->output, 0);
+
+	info("Writing file '%s'...", stream->path.array);
```
Writing file -> Writing to path

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -525,14 +527,19 @@ static int ffmpeg_mux_init_context(struct ffmpeg_mux *ffm)
+	bool isHLS = false;
 	if (strncmp(ffm->params.file, SRT_PROTO, sizeof(SRT_PROTO) - 1) == 0 ||
 	    strncmp(ffm->params.file, UDP_PROTO, sizeof(UDP_PROTO) - 1) == 0 ||
 	    strncmp(ffm->params.file, TCP_PROTO, sizeof(TCP_PROTO) - 1) == 0)
```
Add a check for HTTP protocol here as well.

ushadow replied:
Then if isNetwork is true, call avformat_network_init().

If isNetwork is true and format is not set, guess format to be "mpegts", else if format is set, guess format to be format.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -229,6 +236,8 @@ static void add_muxer_params(struct dstr *cmd, struct ffmpeg_muxer *stream)
 			       const char *path)
 {
+	if ((stream->format == NULL) || (*stream->format == '\0'))
+		stream->format = "";
```
I don't see where stream->format is set in the command line.

mv2731 committed:
plugins: Restructured code logic in ffmpeg-mux.c

ushadow reviewed (changes requested):
Please see the comments.

mv2731 committed:
plugins: Eliminate format parameter

Determine format parameter unnecessary because output_format
can be guessed simply based off the filename. Eliminate
the format parameter across multiple files and fix code logic
in ffmpeg_mux_init_context.

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -256,9 +256,7 @@ static bool init_params(int *argc, char ***argv, struct main_params *params,
 	}
 
 	*p_audio = audio;
-
```
Don't remove the empty line here and the one below.

mv2731 replied:
Done.

mv2731 committed:
plugins: Maintain earlier spacing for ffmpeg-mux.c

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 	AVOutputFormat *output_format;
 	int ret;
 	bool isNetwork = false;
+	bool isHTTP = false;
```
bool is_http = (strncmp(ffm->params.file, HTTP_PROTO,
			  sizeof(HTTP_PROTO) - 1) == 0);

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 {
 	AVOutputFormat *output_format;
 	int ret;
 	bool isNetwork = false;
```
isNetwork -> is_network

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
+	}
+
+	if (isNetwork && !(isHTTP))
 		output_format = av_guess_format("mpegts", NULL, "video/M2PT");
```
!is_http

mv2731 committed:
plugins: Change boolean names

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
+	bool is_network = false;
+	bool is_http = false;
+	is_http = (strncmp(ffm->params.file, HTTP_PROTO,
+			  sizeof(HTTP_PROTO) - 1) == 0);
```
Add a space before sizeof to align with ffm.

mv2731 committed:
plugins: Apply clang-format to ffmpeg-mux.c

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -354,6 +360,50 @@ static bool ffmpeg_mux_start(void *data)
+	obs_output_begin_data_capture(stream->output, 0);
+
+	info("Writing to path '%s'...", stream->path.array);
+	dstr_free(&path);
```
Move dstr_free(&path); to after start_pipe(stream, path.array);

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -354,6 +360,50 @@ static bool ffmpeg_mux_start(void *data)
+	if (!obs_output_initialize_encoders(stream->output, 0))
+		return false;
+
+	settings = obs_output_get_settings(stream->output);
```
We don't really need settings.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -354,6 +360,50 @@ static bool ffmpeg_mux_start(void *data)
+		return false;
+
+	settings = obs_output_get_settings(stream->output);
+	obs_service_t *service;
```
Move obs_service_t *service; to the beginning where all the variables are declared.

mv2731 committed:
plugins: Remove settings var in hls_mux_start function

The settings variable hanging around in hls_mux_start function
is unnecessary because it is never used.

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -519,29 +519,37 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 			ffm->params.file);
 		return FFM_ERROR;
 	}
+	printf("info: output_format name and long_name: %s, %s\n",
```
Check if output_format->name or output_format->long_name is NULL. You can do something like:
output_format->name ? output_format->name : "unknown"

ushadow reviewed (changes requested):
One more small change.

mv2731 committed:
plugins: Added a check for output_format names null

ushadow commented on plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c:
```diff
@@ -519,29 +519,39 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 			ffm->params.file);
 		return FFM_ERROR;
 	}
+	output_format->name ? output_format->name : "unknown";
```
These need to be in the printf statement.

printf("info: output_format name and long_name: %s, %s\n",
           output_format->name ? output_format->name : "unknown",
           output_format->long_name ? output_format->long_name : "unknown");

ushadow reviewed (changes requested):
Please see the comment.

mv2731 committed:
plugins: Edit print log of output_format name
